File: crates/ruimte_runtime/src/processes/tree.rs
use std::collections::{HashMap, HashSet};

use serde_json::{Value, json};

use crate::sessions::ProcessRoot;

use super::sampler::{ProcessRate, RawProcess, RawSample, identity};

const AI_FAMILIES: &[&str] = &[
    "claude", "codex", "gemini", "copilot", "opencode", "ollama", "cursor", "aider",
];

pub fn groups(
    sample: &RawSample,
    rates: &HashMap<(u32, u64), ProcessRate>,
    roots: &[ProcessRoot],
    daemon_pid: u32,
    scope: &str,
    sort: &str,
) -> Vec<Value> {
    let by_pid = sample
        .processes
        .iter()
        .map(|process| (process.pid, process))
        .collect::<HashMap<_, _>>();
    let mut children: HashMap<u32, Vec<&RawProcess>> = HashMap::new();
    for process in &sample.processes {
        children.entry(process.ppid).or_default().push(process);
    }
    for siblings in children.values_mut() {
        siblings.sort_by_key(|process| process.pid);
    }
    let mut claimed = HashSet::new();
    let mut result = Vec::new();
    let node_pids = roots
        .iter()
        .filter(|root| !root.exited)
        .map(|root| root.pid)
        .collect::<HashSet<_>>();
    for root in roots.iter().filter(|root| !root.exited) {
        if let Some(process) = by_pid.get(&root.pid) {
            let entries = walk(process, &children, 0, &mut claimed, None, &HashSet::new());
            result.push(group_json(
                format!("terminal:{}", root.id),
                "terminal",
                Some(&root.id),
                entries,
                rates,
                sort,
                0,
            ));
        }
    }
    if let Some(process) = by_pid.get(&daemon_pid) {
        let entries = walk(process, &children, 0, &mut claimed, None, &node_pids);
        result.push(group_json(
            "daemon".into(),
            "daemon",
            None,
            entries,
            rates,
            sort,
            0,
        ));
    }
    if scope == "all" {
        let mut remaining = sample
            .processes
            .iter()
            .filter(|process| !claimed.contains(&identity(process)))
            .map(|process| Entry {
                process,
                depth: 0,
                family: own_family(process, &[]),
            })
            .collect::<Vec<_>>();
        let hidden = remaining.len().saturating_sub(50);
        remaining.sort_by(|left, right| {
            rate_value(rates.get(&identity(right.process)), sort)
                .total_cmp(&rate_value(rates.get(&identity(left.process)), sort))
        });
        remaining.truncate(50);
        result.push(group_json(
            "other".into(),
            "other",
            None,
            remaining,
            rates,
            sort,
            hidden,
        ));
    }
    result
}

struct Entry<'a> {
    process: &'a RawProcess,
    depth: usize,
    family: Option<String>,
}

fn walk<'a>(
    root: &'a RawProcess,
    children: &HashMap<u32, Vec<&'a RawProcess>>,
    depth: usize,
    claimed: &mut HashSet<(u32, u64)>,
    inherited: Option<String>,
    excluded: &HashSet<u32>,
) -> Vec<Entry<'a>> {
    let family = own_family(root, &[]).or(inherited);
    claimed.insert(identity(root));
    let mut entries = vec![Entry {
        process: root,
        depth,
        family: family.clone(),
    }];
    for child in children.get(&root.pid).into_iter().flatten() {
        if !excluded.contains(&child.pid) {
            entries.extend(walk(
                child,
                children,
                depth + 1,
                claimed,
                family.clone(),
                excluded,
            ));
        }
    }
    entries
}

pub(super) fn own_family(process: &RawProcess, args: &[String]) -> Option<String> {
    let haystack = format!(
        "{} {} {}",
        process.name,
        process.path.as_deref().unwrap_or_default(),
        args.iter().take(2).cloned().collect::<Vec<_>>().join(" ")
    )
    .to_ascii_lowercase();
    AI_FAMILIES
        .iter()
        .find(|family| {
            haystack
                .split(|character: char| "/ ._@-".contains(character))
                .any(|word| word == **family)
        })
        .map(|family| (*family).to_owned())
}

fn group_json(
    id: String,
    kind: &str,
    node_id: Option<&str>,
    mut entries: Vec<Entry<'_>>,
    rates: &HashMap<(u32, u64), ProcessRate>,
    sort: &str,
    hidden: usize,
) -> Value {
    entries.sort_by(|left, right| {
        left.depth.cmp(&right.depth).then_with(|| {
            rate_value(rates.get(&identity(right.process)), sort)
                .total_cmp(&rate_value(rates.get(&identity(left.process)), sort))
        })
    });
    let readable = entries
        .iter()
        .filter_map(|entry| rates.get(&identity(entry.process)))
        .collect::<Vec<_>>();
    let sum = |field: fn(&ProcessRate) -> Option<f64>| {
        let values = readable
            .iter()
            .filter_map(|rate| field(rate))
            .collect::<Vec<_>>();
        (!values.is_empty()).then(|| values.into_iter().sum::<f64>())
    };
    let memory = readable
        .iter()
        .filter_map(|rate| rate.memory)
        .collect::<Vec<_>>();
    json!({
        "id": id,
        "kind": kind,
        "nodeId": node_id,
        "cpu": sum(|rate| rate.cpu),
        "memory": (!memory.is_empty()).then(|| memory.into_iter().sum::<u64>()),
        "diskRead": sum(|rate| rate.disk_read),
        "diskWrite": sum(|rate| rate.disk_write),
        "processes": entries.into_iter().map(|entry| row_json(entry, rates)).collect::<Vec<_>>(),
        "hidden": hidden,
    })
}

fn row_json(entry: Entry<'_>, rates: &HashMap<(u32, u64), ProcessRate>) -> Value {
    let process = entry.process;
    let rate = rates.get(&identity(process));
    json!({
        "pid": process.pid,
        "startTime": process.start_time,
        "ppid": process.ppid,
        "name": process.name,
        "path": process.path,
        "readable": process.readable,
        "cpu": rate.and_then(|rate| rate.cpu),
        "memory": rate.and_then(|rate| rate.memory),
        "diskRead": rate.and_then(|rate| rate.disk_read),
        "diskWrite": rate.and_then(|rate| rate.disk_write),
        "family": entry.family,
        "depth": entry.depth,
    })
}

fn rate_value(rate: Option<&ProcessRate>, sort: &str) -> f64 {
    match sort {
        "memory" => rate.and_then(|rate| rate.memory).unwrap_or(0) as f64,
        "disk" => rate
            .map(|rate| rate.disk_read.unwrap_or(0.) + rate.disk_write.unwrap_or(0.))
            .unwrap_or(0.),
        _ => rate.and_then(|rate| rate.cpu).unwrap_or(0.),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn process(pid: u32, ppid: u32, name: &str) -> RawProcess {
        RawProcess {
            pid,
            ppid,
            uid: 501,
            start_time: pid as u64,
            name: name.into(),
            path: None,
            readable: true,
            cpu_ns: None,
            memory: None,
            disk_read: None,
            disk_write: None,
        }
    }

    #[test]
    fn session_processes_are_not_duplicated_under_the_daemon() {
        let sample = RawSample {
            at: 0,
            awake_ms: 0,
            asleep_ms: 0,
            processes: vec![
                process(10, 1, "daemon"),
                process(20, 10, "shell"),
                process(21, 20, "claude"),
            ],
            machine: super::super::sampler::MachineCounters {
                cores: 1,
                cpu_busy: None,
                cpu_total: None,
                memory_used: None,
                memory_total: 0,
                disk_free: None,
                disk_total: None,
            },
        };
        let roots = vec![ProcessRoot {
            id: "terminal".into(),
            pid: 20,
            exited: false,
        }];

        let result = groups(&sample, &HashMap::new(), &roots, 10, "ruimte", "cpu");

        assert_eq!(result[0]["processes"].as_array().unwrap().len(), 2);
        assert_eq!(result[1]["processes"].as_array().unwrap().len(), 1);
        assert_eq!(result[0]["processes"][1]["family"], "claude");
        assert!(result[0]["cpu"].is_null());
    }
}

File: apps/server/src/processes/tree.ts
import type { ProcessGroup, ProcessGroupKind, ProcessRow, ProcessScope, ProcessSort } from '@ruimte/contracts';
import { identityOf, type ProcessRate, type RawProcess, type RawSample } from './sampler.ts';

/* The AI CLIs and apps recognized by name or path, so a `claude` started from iTerm stands out in "All". */
export const AI_FAMILIES = ['claude', 'codex', 'gemini', 'copilot', 'opencode', 'ollama', 'cursor', 'aider'] as const;

// A whole word of the path or name: `~/.local/share/claude/versions/2.1.269`, `Claude.app`, `codex-aarch64`.
const FAMILY_PATTERNS = AI_FAMILIES.map((family) => [family, new RegExp(`(^|[/\\s._@-])${family}([/\\s._-]|$)`, 'i')] as const);

/* Which AI the process itself is, from its path, its name and the script a runtime was given. */
export const ownFamilyOf = (name: string, path: string | null, args: readonly string[] | null = null): string | null => {
    // A native CLI names itself in its path; one on node or bun names itself in the script it runs.
    const haystack = [path ?? '', name, ...(args?.slice(0, 2) ?? [])].join(' ');
    for (const [family, pattern] of FAMILY_PATTERNS) {
        if (pattern.test(haystack)) {
            return family;
        }
    }
    return null;
};

export interface TreeRoots {
    daemonPid: number;
    sessions: readonly { id: string; pid: number }[];
    chats: readonly { id: string; pid: number }[];
}

export interface TreeEntry {
    process: RawProcess;
    identity: string;
    depth: number;
}

export interface IndexedGroup {
    id: string;
    kind: ProcessGroupKind;
    nodeId: string | null;
    entries: TreeEntry[];
}

export interface TreeIndex {
    sample: RawSample;
    byPid: Map<number, RawProcess>;
    /* Every group of Ruimte's own tree, in a stable order. */
    groups: IndexedGroup[];
    /* The identities inside Ruimte's tree. */
    ruimte: Set<string>;
    /* Inherited: a shell under `claude` is `claude` too. */
    family: Map<string, string | null>;
    /* The process's own, which is what "the agent is still there" asks about. */
    ownFamily: Map<string, string | null>;
}

/* Electron's main process, or the app bundle a packaged daemon runs in. */
const looksLikeDesktopApp = (process: RawProcess): boolean => /electron|\.app\/Contents\/MacOS\//i.test(process.path ?? process.name);

/*
 * Sorts the process table into Ruimte's groups: a node per terminal and chat, the desktop app when
 * the daemon runs inside it, and the daemon with whatever else it started (probes, git). A family is
 * inherited up to the root of a group and never past it, so a daemon started from a Claude session
 * does not make every shell on its canvas a Claude.
 */
export const indexTree = (sample: RawSample, roots: TreeRoots, argsOf: (process: RawProcess) => readonly string[] | null = () => null): TreeIndex => {
    const byPid = new Map<number, RawProcess>();
    const children = new Map<number, RawProcess[]>();
    for (const process of sample.processes) {
        byPid.set(process.pid, process);
    }
    for (const process of sample.processes) {
        if (process.ppid === process.pid || !byPid.has(process.ppid)) {
            continue;
        }
        const siblings = children.get(process.ppid);
        if (siblings) {
            siblings.push(process);
        } else {
            children.set(process.ppid, [process]);
        }
    }
    for (const siblings of children.values()) {
        siblings.sort((a, b) => a.pid - b.pid);
    }

    const nodeRoots = new Map<number, { id: string; kind: ProcessGroupKind; nodeId: string }>();
    for (const session of roots.sessions) {
        nodeRoots.set(session.pid, { id: `terminal:${session.id}`, kind: 'terminal', nodeId: session.id });
    }
    for (const chat of roots.chats) {
        nodeRoots.set(chat.pid, { id: `chat:${chat.id}`, kind: 'chat', nodeId: chat.id });
    }

    const walk = (root: RawProcess, skip: (process: RawProcess) => boolean): TreeEntry[] => {
        const entries: TreeEntry[] = [];
        const visit = (process: RawProcess, depth: number): void => {
            entries.push({ process, identity: identityOf(process.pid, process.startTime), depth });
            for (const child of children.get(process.pid) ?? []) {
                if (!skip(child)) {
                    visit(child, depth + 1);
                }
            }
        };
        visit(root, 0);
        return entries;
    };

    const groups: IndexedGroup[] = [];
    for (const [pid, root] of nodeRoots) {
        const process = byPid.get(pid);
        if (process) {
            groups.push({ ...root, entries: walk(process, () => false) });
        }
    }
    const daemon = byPid.get(roots.daemonPid);
    const parent = daemon ? byPid.get(daemon.ppid) : undefined;
    if (parent && parent.pid > 1 && looksLikeDesktopApp(parent)) {
        groups.push({ id: 'app', kind: 'app', nodeId: null, entries: walk(parent, (process) => process.pid === roots.daemonPid) });
    }
    if (daemon) {
        groups.push({ id: 'daemon', kind: 'daemon', nodeId: null, entries: walk(daemon, (process) => nodeRoots.has(process.pid)) });
    }

    const ruimte = new Set<string>();
    const boundaries = new Set<number>([roots.daemonPid, ...nodeRoots.keys()]);
    const ownFamily = new Map<string, string | null>();
    for (const group of groups) {
        for (const entry of group.entries) {
            ruimte.add(entry.identity);
        }
    }
    const own = (process: RawProcess, identity: string): string | null => {
        let family = ownFamily.get(identity);
        if (family === undefined) {
            family = ownFamilyOf(process.name, process.path, ruimte.has(identity) ? argsOf(process) : null);
            ownFamily.set(identity, family);
        }
        return family;
    };
    const family = new Map<string, string | null>();
    const inherited = (process: RawProcess): string | null => {
        const chain: RawProcess[] = [];
        let found: string | null = null;
        let current: RawProcess | undefined = process;
        // Bounded, because a table read while processes exit can hold a stale parent that loops.
        while (current !== undefined && chain.length < 64) {
            const identity = identityOf(current.pid, current.startTime);
            const known = family.get(identity);
            if (known !== undefined) {
                found = known;
                break;
            }
            chain.push(current);
            found = own(current, identity);
            if (found !== null || boundaries.has(current.pid) || current.pid <= 1) {
                break;
            }
            current = byPid.get(current.ppid);
        }
        for (const link of chain) {
            family.set(identityOf(link.pid, link.startTime), found);
        }
        return found;
    };
    for (const process of sample.processes) {
        inherited(process);
    }
    return { sample, byPid, groups, ruimte, family, ownFamily };
};

const sortValue = (rate: Pick<ProcessRate, 'cpu' | 'memory' | 'diskRead' | 'diskWrite'> | undefined, sort: ProcessSort): number => {
    if (rate === undefined) {
        return -1;
    }
    if (sort === 'cpu') {
        return rate.cpu ?? -1;
    }
    if (sort === 'memory') {
        return rate.memory ?? -1;
    }
    return rate.diskRead === null && rate.diskWrite === null ? -1 : (rate.diskRead ?? 0) + (rate.diskWrite ?? 0);
};

const sum = (values: readonly (number | null)[]): number | null => {
    let total: number | null = null;
    for (const value of values) {
        if (value !== null) {
            total = (total ?? 0) + value;
        }
    }
    return total;
};

const rowOf = (index: TreeIndex, rates: Map<string, ProcessRate>, entry: TreeEntry): ProcessRow => {
    const rate = rates.get(entry.identity);
    const { process } = entry;
    return {
        pid: process.pid,
        startTime: process.startTime,
        ppid: process.ppid,
        name: process.name,
        path: process.path,
        readable: process.readable,
        cpu: rate?.cpu ?? null,
        memory: rate?.memory ?? null,
        diskRead: rate?.diskRead ?? null,
        diskWrite: rate?.diskWrite ?? null,
        family: index.family.get(entry.identity) ?? null,
        depth: entry.depth
    };
};

const groupOf = (id: string, kind: ProcessGroupKind, nodeId: string | null, rows: ProcessRow[], hidden = 0): ProcessGroup => ({
    id,
    kind,
    nodeId,
    cpu: sum(rows.map((row) => row.cpu)),
    memory: sum(rows.map((row) => row.memory)),
    diskRead: sum(rows.map((row) => row.diskRead)),
    diskWrite: sum(rows.map((row) => row.diskWrite)),
    processes: rows,
    hidden
});

const GROUP_ORDER: Record<ProcessGroupKind, number> = { terminal: 0, chat: 0, app: 1, daemon: 2, other: 3 };

/*
 * What the panel draws for one scope. "All" adds the rest of the machine, but only the top of it on
 * the chosen sort plus every AI process, never 1,500 rows every two seconds.
 */
export const groupsFor = (index: TreeIndex, rates: Map<string, ProcessRate>, scope: ProcessScope, sort: ProcessSort, limit: number): ProcessGroup[] => {
    const groups = index.groups.map((group) =>
        groupOf(
            group.id,
            group.kind,
            group.nodeId,
            group.entries.map((entry) => rowOf(index, rates, entry))
        )
    );
    groups.sort((a, b) => GROUP_ORDER[a.kind] - GROUP_ORDER[b.kind] || (GROUP_ORDER[a.kind] === 0 ? sortValue(b, sort) - sortValue(a, sort) : 0));
    if (scope === 'all') {
        const rest = index.sample.processes
            .map((process) => ({ process, identity: identityOf(process.pid, process.startTime), depth: 0 }))
            .filter((entry) => entry.process.pid > 0 && !index.ruimte.has(entry.identity));
        rest.sort((a, b) => sortValue(rates.get(b.identity), sort) - sortValue(rates.get(a.identity), sort));
        const shown = rest.filter((entry, position) => (position < limit && entry.process.readable) || (index.ownFamily.get(entry.identity) ?? null) !== null);
        groups.push(
            groupOf(
                'other',
                'other',
                null,
                shown.map((entry) => rowOf(index, rates, entry)),
                rest.length - shown.length
            )
        );
    }
    return groups;
};

/* The share of Ruimte for the charts: percent of one core summed, bytes, bytes per second. */
export const ruimteTotals = (index: TreeIndex, rates: Map<string, ProcessRate>): { cpu: number | null; memory: number | null; disk: number | null } => {
    const members = [...index.ruimte].map((identity) => rates.get(identity)).filter((rate) => rate !== undefined);
    return {
        cpu: sum(members.map((rate) => rate.cpu)),
        memory: sum(members.map((rate) => rate.memory)),
        disk: sum(members.map((rate) => (rate.diskRead === null && rate.diskWrite === null ? null : (rate.diskRead ?? 0) + (rate.diskWrite ?? 0))))
    };
};

/* Disk throughput of the whole machine as far as it can be read, the sum of every readable process. */
export const machineDisk = (rates: Map<string, ProcessRate>): { read: number | null; write: number | null } => {
    const all = [...rates.values()];
    return { read: sum(all.map((rate) => rate.diskRead)), write: sum(all.map((rate) => rate.diskWrite)) };
};
